contracts/AgentMech.sol

```diff
@@ -22,19 +22,26 @@ error AgentNotFound(uint256 agentId);
 /// @param expected Expected amount.
 error NotEnoughPaid(uint256 provided, uint256 expected);
 
+/// @dev Request Id not found.
+/// @param requestId Request Id.
+error RequestIdNotFound(uint256 requestId);
+
 /// @title AgentMech - Smart contract for extending ERC721Mech
 /// @dev A Mech that is operated by the holder of an ERC721 non-fungible token.
 contract AgentMech is ERC721Mech {
-    event Perform(address indexed sender, bytes32 taskHash);
     event Deliver(address indexed sender, uint256 requestId, bytes data);
     event Request(address indexed sender, uint256 requestId, bytes data);
     event PriceUpdated(uint256 price);
 
     // Minimum required price
     uint256 public price;
+    // Number of undelivered requests
+    uint256 public numUndeliveredRequests;
 
     // Map of requests counts for corresponding addresses
     mapping (address => uint256) public mapRequestsCounts;
+    // Map of request Ids
+    mapping (uint256 => uint256[2]) public mapRequestIds;
```

Comment:
A two-way circular map stores all the undelivered requests and correctly removes elements when requests are delivered.